src/main/kotlin/cacophony/regex/RegexParser.kt

```diff
@@ -0,0 +1,148 @@
+package cacophony.regex
+
+import cacophony.utils.AlgebraicRegex
+import java.util.EmptyStackException
+import java.util.Stack
+
+fun parseRegex(str: String): AlgebraicRegex {
+    val parser = RegexParser()
+    val regex = "($str)"
+    var specialCharacter = false
+    var lastEscaped = -1
+    regex.indices.forEach {
+        val c = regex[it]
+        if (specialCharacter) {
+            if (c !in SPECIAL_CHARACTER_MAP.keys) throw RegexSyntaxErrorException()
```

Comment:
Maybe we can add more descriptive errors? Like `throw RegexSyntaxErrorException("Expected special character, but got $c at position x")`. Similarly in other places.